modeshape-jcr/src/main/java/org/modeshape/jcr/shell/ShellConsole.java

```diff
@@ -0,0 +1,223 @@
+        if (cmd.equals("list")) {
+            listRepos();
+            return 0;
+        }
+
+        Command command = Commands.lookup(cmd);
+
+        if (command == null) {
+            console.getShell().err().println("Unknown command " + cmd);
+            return 0;
+        }
+
+        command.prepare(cmd);
+
+        if (command.name().equals("connect")) {
+            if (!selectChannel((ConnectCommand) command)) {
+                return 0;
+            }
+        }
+
+        if (command.name().equals("script")) {
+            scriptCommand(command, command.optionValue("-f"));
+            return 0;
+        }
+
+        if (mbeanProxy == null) {
+            console.getShell().err().println("Connect to repository first");
+            return 0;
+        }
+
+        perform(cmd);
+        return 0;
+    }
+
+    private boolean selectChannel(ConnectCommand cmd) {
+        String repoName = cmd.args(0);
+
+        if (repoName == null) {
+            console.getShell().err().println("Repository name expected");
+            return false;
+        }
+
+        String uname = cmd.optionValue("-u");
+        if (uname != null) {
+            user = uname;
+        }
+
+        try {
+            ObjectName mbeanName = new ObjectName("org.modeshape:type=" + repoName);
+            console.setPrompt(prompt(String.format("[%s@%s]$ ", user, repoName)));
+
+            mbeanProxy =
+                    JMX.newMBeanProxy(mbsc, mbeanName, ShellMBean.class, true);
+            return true;
+        } catch (Exception e) {
+            console.getShell().err().println(e.getMessage());
+            return false;
+        }
+    }
+
+    private void perform(String cmd) throws InterruptedException {
+        try {
+            String res = mbeanProxy.perform(cmd);
+            console.getShell().out().println(res);
+        } catch (Exception e) {
+            console.getShell().err().println("" + e.getMessage());
+        }
+    }
+
+    private Prompt prompt(String prompt) {
+        List<TerminalCharacter> chars = new ArrayList<>();
+        for (int i = 0; i < prompt.length(); i++) {
+            chars.add(new TerminalCharacter(prompt.charAt(i), new TerminalColor(Color.DEFAULT, Color.DEFAULT)));
+        }
+        return new Prompt(chars);
+    }
+
+    private void listRepos() {
+        try {
+            ObjectName n = new ObjectName("org.modeshape:type=*");
+            Set<ObjectName> names = mbsc.queryNames(n, null);
+            for (ObjectName nn : names) {
+                console.getShell().out().println(nn.getKeyProperty("type"));
+            }
+        } catch (MalformedObjectNameException | IOException e) {
+            console.getShell().err().println(e.getMessage());
+        }
+    }
+
+    private void scriptCommand(Command command, String fileName) {
+        if (fileName == null) {
+            console.getShell().err().println(command.help());
+            return;
+        }
+
+        try {
```

Comment:
wouldn't a `try-with-resources` be better here ? It doesn't look like the streams are closed anywhere